cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataServiceModule.java

```diff
@@ -17,25 +17,70 @@
 package co.cask.cdap.metadata;
 
 import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.runtime.RuntimeModule;
 import co.cask.cdap.gateway.handlers.CommonHandlers;
+import co.cask.cdap.proto.metadata.MetadataChangeRecord;
 import co.cask.http.HttpHandler;
 import com.google.inject.AbstractModule;
+import com.google.inject.Module;
 import com.google.inject.multibindings.Multibinder;
 import com.google.inject.name.Names;
 
 /**
  * Guice module for metadata service.
  */
-public class MetadataServiceModule extends AbstractModule {
+public class MetadataServiceModule extends RuntimeModule {
+
+  @Override
+  public Module getInMemoryModules() {
+    return new InMemoryMetadataModule();
+  }
+
   @Override
-  protected void configure() {
-    Multibinder<HttpHandler> handlerBinder = Multibinder.newSetBinder(
-      binder(), HttpHandler.class, Names.named(Constants.Metadata.METADATA_HANDLERS_NAME));
+  public Module getStandaloneModules() {
+    return new InMemoryMetadataModule();
+  }
+
+  @Override
+  public Module getDistributedModules() {
+    return new DistributedMetadataModule();
+  }
+
+  private static final class CommonMetadataModule extends AbstractModule {
+    @Override
+    protected void configure() {
+      Multibinder<HttpHandler> handlerBinder = Multibinder.newSetBinder(
+        binder(), HttpHandler.class, Names.named(Constants.Metadata.METADATA_HANDLERS_NAME));
+
+      CommonHandlers.add(handlerBinder);
+      handlerBinder.addBinding().to(MetadataHttpHandler.class);
+      handlerBinder.addBinding().to(LineageHandler.class);
+
+      bind(MetadataAdmin.class).to(DefaultMetadataAdmin.class);
+    }
+  }
+
+  private static final class DistributedMetadataModule extends AbstractModule {
+    @Override
+    protected void configure() {
+      install(new CommonMetadataModule());
```

Comment:
Better not to use install. You can extend from the CommonModule and implements a abstract method defined in CommonModule.